### DISCUSSION TITLE
Proposal to populate path_parameters in Request before mddleware execution

### DISCUSSION THREAD

**Discussion by Matteo-Frattaroli** [fastapi/fastapi#12957] [Questions] [answered]
Hello everyone,

While working on a FastAPI application, I encountered a scenario where I needed to automatically log path parameters for specific routes extracting them and setting a context var used by a custom logger. To achieve this, I wrote a middleware to extract path parameters during each request. However, even though I knew that that middlewares are executed before the routing phase, I also noticed a request.path_parameters object was present but that it's empty during middleware execution.

Although this is the expected lifecycle in FastAPI, I found that counterintuitive since all the necessary elements to populate path_parameters are already available in the middleware context, namely:
- The app object, which includes all registered routers with their unformatted route templates (e.g., my/route/{my_param}/stuff).
- The fully formatted path (e.g., request.url as my/route/myPathParamValue1/stuff).

To work around this, I implemented a custom solution by matching the incoming request path against the route definitions using regex. While this works, it feels redundant since FastAPI already has robust regex-based matching in Router and APIRoute.

**Proposal**
I propose enhancing FastAPI to populate request.path_parameters (or scope["path_parameters"]) before middleware execution, making path parameters accessible throughout the entire request lifecycle. This would:
•	Simplify middleware development for use cases like logging, monitoring, or custom authentication.
- Eliminate the need for developers to reimplement route-matching logic or force them to access and match nested objects in middleware.
- Align with the expectation that Request contains all relevant metadata from the start.

Do you think this feature align with Starlette’s lifecycle design, or should it remain as a middleware-level customization? If so is it maybe worth it to add a special middleware class for such cases?

If the community finds this idea valuable, I propose:
- Modifying the routing process (e.g., in Router.__call__) to populate scope["path_parameters"] once a match is found.
- Exposing path_parameters as a native attribute of Request (e.g., request.path_parameters).
- Ensuring backward compatibility, potentially making the feature opt-in via a configuration flag (e.g., enable_path_parameters_in_scope).

I’d be happy to contribute a pull request with the actual implementation, tests and docs.

Looking forward to hearing your feedback!

**Comment by luzzodev** [comment, marked as answer]
Hi @Matteo-Frattaroli,

here some considerations about yout proposal:

Middleware can be used to manipulate request/response in general and part of this changes can affect/influence the routing phase itself, so moving routing before middleware execution is not a good moves.

I saw several people dealing with logging and path parameter/route information and this can be easily achived by using a middleware and read the routing info by intercepting the response.

A cool and usefull feature could be to have an additional layer of middleware execution after the routing phase but this another and more complex topic due to the implications it could have.

About other system like auth etc, you can go with dependency at different level or write a simple decorator.

A final consideration: having a reach ecosystem of middleware could be really valuable and positive thing but they should't be attached to the Fastapi main class via a simple param.
